proposal/embedded-protocol-2.md

```diff
@@ -0,0 +1,168 @@
+
+## Background
+
+> This section is non-normative.
+
+Since the [embedded Sass protocol] was first released, a number of improvements
+have been suggested by embedded host authors that would either be
+backwards-incompatible, or would require cumbersome redundancies in the protocol
+to make them backwards-compatible.
+
+[embedded Sass protocol]: ../spec/embedded-protocol.md
+
+In the meantime, to provide an efficient Sass CLI along with embedded hosts, the
+Dart Sass embedded compiler was merged into the main Dart Sass executable.
+Because this is effectively also a breaking change in how hosts obtain and
+invoke the compiler, it seemed like a good opportunity to make the other
+breaking protocol changes that had been piling up.
+
+## Summary
+
+> This section is non-normative.
+
+This proposal makes three breaking changes to the embedded Sass protocol:
+
+* Replace the compilation IDs in the protobufs with one directly encoded in the
+  wire protocol so that compiler implementations can efficiently route requests
+  to different workers without reparsing messages.
+
+* Use optional fields as defined in protocol buffers 3.15.0 instead of treating
+  default field values as absent.
+
+* Move `CompileSuccess.loaded_urls` to `CompileResult.loaded_urls` so it's
+  available even when compilation fails.
+
+## Overview
+
+Replace the last paragraph of the [embedded protocol overview] with:
+
+[embedded protocol overview]: ../spec/embedded-protocol.md#overview
+
+Each message in the embedded protocol is sent as a _packet_ which contains two
+values: an unsigned [varint] up to 32 bits long known as the "compilation ID",
+and a protocol buffer that contains the protobuf message. For streams (like
+standard input and output) that don't have built-in message boundaries, every
+packet must begin with another unsigned varint indicating the length in bytes of
+the remaining message (_including the compilation ID_). This matches the best
+practice described in [the protocol buffer documentation].
+
+Because JavaScript can't easily represent integers larger than 2^53 - 1, the
+length may be no more than 2^53 - 1. Because it's so unlikely that this will
+come up in practice, implementations are not required to verify it.
+
+[varint]: https://developers.google.com/protocol-buffers/docs/encoding#varints
+[the protocol buffer documentation]: https://developers.google.com/protocol-buffers/docs/techniques#streaming
+
+For a length-delimited stream, each packet has the following structure:
+
+```
+╔══════════╦══════════════════╗
+║ varint   ║ Length           ║
+╠══════════╬══════════════════╣
+║ varint   ║ Compilation ID   ║
+╠══════════╬══════════════════╣
+║ protobuf ║ Protobuf Message ║
+╚══════════╩══════════════════╝
+```
+
+## RPCs
+
+### Type Definitions
+
+Replace the following RPC type definitions:
+
+* *Requests* usually include a mandatory `uint32 id` field so that the other
+  endpoint can respond, except for `CompilationRequest` which uses the
```

Comment:
The text on these both bullet points mention that `Compile{Request,Response}` use the compilation ID as its ID. However these fields will be removed from the proto message. I get that this ID is being bubbled up to the wire packet, but these paragraphs sound like they refer to the proto message.

The paragraphs also start saying that "usually" messages will include a mandatory `id`. I wonder if maybe these messages would be better worded as something like:

> - *Requests* must include a mandatory `uint32 id`, except for `CompileRequest` which uses the wire packet compilation ID as its own ID.

(or something better, but along those lines 🤔)

Reply:
For compiler side requests like FunctionCallRequest etc. In theory, it is possible to have more than 1 requests inflight for the same compilation therefore we need id to track unique requests inside a single compilation.

Reply:
The idea is that this is referring to "compilation ID" as a term-of-art which was defined in the overview. To make that clearer, I'll add a "Packet Structure" subheading and link it from the term.